com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs

```diff
@@ -0,0 +1,440 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Unity.Collections;
+using Unity.Netcode.Transports.UNET;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+namespace Unity.Netcode
+{
+
+    public class InvalidMessageStructureException : SystemException
+    {
+        public InvalidMessageStructureException() { }
+        public InvalidMessageStructureException(string issue) : base(issue) { }
+    }
+    
+    public class MessagingSystem: IMessageHandler, IDisposable
+    {
+#region Internal Types
+        private struct ReceiveQueueItem
+        {
+            public FastBufferReader Reader;
+            public MessageHeader Header;
+            public ulong SenderId;
+            public float Timestamp;
+        }
+
+        private struct SendQueueItem
+        {
+            public BatchHeader BatchHeader;
+            public FastBufferWriter Writer;
+            public readonly NetworkDelivery NetworkDelivery;
+
+            public SendQueueItem(NetworkDelivery delivery, int writerSize, Allocator writerAllocator, int maxWriterSize = -1)
+            {
+                Writer = new FastBufferWriter(writerSize, writerAllocator, maxWriterSize);
+                NetworkDelivery = delivery;
+                BatchHeader = default;
+            }
+        }
+        
+        internal delegate void MessageHandler(ref FastBufferReader reader, NetworkContext context);
+#endregion
+
+#region Private Members
+        private DynamicUnmanagedArray<ReceiveQueueItem> m_IncomingMessageQueue = new DynamicUnmanagedArray<ReceiveQueueItem>(16);
+
+        private MessageHandler[] m_MessageHandlers = new MessageHandler[255];
+        private Type[] m_ReverseTypeMap = new Type[255];
+                
+        private Dictionary<Type, byte> m_MessageTypes = new Dictionary<Type, byte>();
+        private NativeHashMap<ulong, Ref<DynamicUnmanagedArray<SendQueueItem>>> m_SendQueues = new NativeHashMap<ulong, Ref<DynamicUnmanagedArray<SendQueueItem>>>(64, Allocator.Persistent);
```

Comment:
Just as a note, I could replace `Ref<DynamicUnmanagedArray<SendQueueItem>>` with `List<SendQueueItem>` and change `SendQueueItem` to a class instead of a struct and change this from `NativeHashMap` to `Dictionary`, and things will work the way they need to. But it'll result in GC every time a client connects and every time a new batch is started. So that's why I'm doing it this way even though it's a bit more dangerous and C++-like.

Reply:
I'd definitely support using & embracing Unity.Collections usage but I generally have mixed feelings about crafting our own utilities & data structures like `Ref<T>`, `DynamicUnmanagedArray<T>` etc.
can we use more Unity stuff if possible and where it makes sense instead of .NET types?

Reply:
Unity collections don't work here. I specifically need the ability to GetByRef() when interacting with elements of the list. List<> and NativeArray<> both return copies of the items in the list if they're value types. And if I make it not a value type, NativeArray<> becomes not an option - so it's either DynamicUnmanagedArray of structs, or List of class types + gc every time a client connects.

But... clients don't connect that often, so that might be an acceptable amount of gc.

Reply:
Actually I've just learned NativeList wraps UnsafeList, and UnsafeList has `ref T ElementAt`, which is exactly what I need. I'll replace `DynamicUnmanagedArray` with `NativeList`. And since the NativeList is just a wrapper around a pointer to the UnsafeList, I won't need to use Ref<> there either.

```c#
NativeHashMap<ulong, NativeList<SendQueueItem>>
```

supports everything I need. I'll make that change once I get the develop merge finished.